src/qililab/calibration/calibration_controller.py

```diff
@@ -0,0 +1,527 @@
+
+        3. **Low-Level Methods**: ``check_state()``, ``check_data()``, and ``calibrate()`` are the methods you would be calling during this process to interact with the ``nodes``.
+
+    Finally, ``run_automatic_calibration()`` is designed to start acquiring data or calibrating in the optimal location of the graph to avoid extra work:
+
+    - If node A has been calibrated very recently (before the ``drift_timeout`` of the :class:`.CalibrationNode`), it would be a waste of resources to check its data, so ``check_state()`` makes ``maintain()`` skip it.
+    - If node A depends on node B, before calibrating node A, we check the data of node B. Calibrating A would be a waste of resources if we were doing so based on faulty data, so it goes to its dependencies first.
+
+    |
+
+    **Dangerous Behaviors:**
+
+    Note that depending on your ``CalibrationController`` construction, you can have dangerous behaviors in the workflow. You need to watch out for:
+
+    - If you give bad ``comparison_thresholds`` or have bad ``comparison_models``, returning ``out_of_spec`` when you actually have ``bad_data`` or the other way around, will make ``diagnose()`` not being able to work properly, since for it to work, you need to have a single layer node separation (``out_of_spec``) between the ``in_spec`` and the ``bad_data`` nodes.
+    - If you give too long ``drift_timeout``'s, since ``maintain()`` will assume the node is 100% working, you will go to further dependents when maybe you shouldn't without recalibrating. In the end ``diagnose()`` saves the day here, since it doesn't do ``check_state()``, although it will be less efficient.
+
+    .. note:: Find more information about the automatic calibration workflow at https://arxiv.org/abs/1803.03226.
+
+    Args:
+        calibration_graph (nx.DiGraph): The calibration (directed acyclic) graph, where each node is a ``string`` corresponding to a ``CalibrationNode.node_id``. Directions should be given
+            by `nodes` pointing to their next `dependents` (natural time flow for calibration), defining our `starts` and `ends` of the calibration as the `roots` (``in_degree=0``) and `leaves`
+            (``out_degree=0``) of the graph.
+
+        node_sequence (dict[str, CalibrationNode]): Mapping for the nodes of the graph, from strings into the actual initialized nodes.
+        runcard (str): The runcard path, containing the serialized platform where the experiments will be run.
+
+    Examples:
+        This example shows how to create 2 nodes twice, one for each qubit, and pass them to a :class:`.CalibrationController` to run the ``maintain()`` algorithm on the second one.
+
+        .. code-block:: python
+
+            import numpy as np
```

Comment:
Make sure examples are "copy-pasteable". Here we are missing some imports.

Reply:
I'll change it!

Reply:
Solved!